src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java

```diff
@@ -0,0 +1,130 @@
+package uk.gov.companieshouse.api.strikeoffobjections.file;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.multipart.MultipartFile;
+import uk.gov.companieshouse.api.strikeoffobjections.common.ApiLogger;
+
+import java.io.IOException;
+
+public class FileTransferApiClient {
+
+    private static final String HEADER_API_KEY = "x-api-key";
+    private static final String UPLOAD = "upload";
+    private static final String CONTENT_DISPOSITION_VALUE = "form-data; name=%s; filename=%s";
+    private static final String NULL_RESPONSE_MESSAGE = "null response from file transfer api url";
+
+    private String requestId;
+
+    @Autowired
+    private ApiLogger logger;
+
+    @Autowired
+    private RestTemplate restTemplate;
+
+    @Value("${FILE_TRANSFER_API_URL}")
+    private String fileTransferApiURL;
+
+    @Value("${FILE_TRANSFER_API_KEY}")
+    private String fileTransferApiKey;
+
+
+    private <T> FileTransferApiClientResponse makeApiCall(FileTransferOperation <T> operation,
+                                                          FileTransferResponseBuilder <T> responseBuilder) {
+        FileTransferApiClientResponse response = new FileTransferApiClientResponse();
+
+        try {
+            T operationResponse = operation.execute();
+            response = responseBuilder.createResponse(operationResponse);
+        } catch (IOException e) {
+            logger.errorContext(requestId, e.getMessage(), e);
+            response.setHttpStatus(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+        return response;
+    }
+
+    /**
+     * Uploads a file to the file-transfer-api
+     * Creates a multipart form request containing the file and sends to
+     * the file-transfer-api. The response from the file-transfer-api contains
+     * the new unique id for the file. This is captured and returned in the FileTransferApiClientResponse.
+     * @param fileToUpload The file to upload
+     * @return FileTransferApiClientResponse containing the file id if successful, the http header and http status
+     */
+    public FileTransferApiClientResponse upload(String requestId, MultipartFile fileToUpload) {
+        this.requestId = requestId;
```

Comment:
Can we guarentee that only one request will be being executed by this object at once?
As if multiple requests go through the object at the same time, then all their request id's will overwrite and interfere with each other.

If each new request instantiates a new FileTransferApiClient that will be ok.

Though would it be better if FileTransferApiClient was a Component that was instantiated by spring and was autowired into services that need it? In which case I would suggest passing the requestId in as method arguments where needed to avoid having "state"

Reply:
Fair point - it is risky  so removed this.